base_layer/mmr/src/common.rs

```diff
@@ -147,47 +150,42 @@ pub fn peak_map_height(mut pos: usize) -> (usize, usize) {
     (bitmap, pos)
 }
 
-/// sizes of peaks and height of next node in mmr of given size
-/// Example: on input 5 returns ([3,1], 1) as mmr state before adding 5 was
+/// Is the node at this pos the "left" sibling of its parent?
+pub fn is_left_sibling(pos: usize) -> bool {
+    let (peak_map, height) = peak_map_height(pos);
+    let peak = 1 << height;
+    (peak_map & peak) == 0
+}
+
+pub fn hash_together<D: Digest>(left: &[u8], right: &[u8]) -> Hash {
+    D::new().chain(left).chain(right).result().to_vec()
+}
+
+/// The number of leaves in a MMR of the provided size.
+/// Example: on input 5 returns (2 + 1 + 1) as mmr state before adding 5 was
 ///    2
 ///   / \
 ///  0   1   3   4
-pub fn peak_sizes_height(size: usize) -> (Vec<usize>, usize) {
+/// None is returned if the number of leaves exceeds the maximum value of a usize
+pub fn checked_n_leaves(size: usize) -> Option<usize> {
     if size == 0 {
-        return (vec![], 0);
+        return Some(0);
     }
     let mut peak_size = ALL_ONES >> size.leading_zeros();
-    let mut sizes = vec![];
+    let mut nleaves = 0usize;
     let mut size_left = size;
     while peak_size != 0 {
         if size_left >= peak_size {
-            sizes.push(peak_size);
+            nleaves += peak_size.checked_add(1)? / 2;
```

Comment:
If you really want to push performance, `n/2` ==> `n >> 1`

Reply:
I 100% do want to push performance 🏎️  - thanks